### DISCUSSION TITLE
send text to pane

### DISCUSSION THREAD

**Discussion by russarr** [wezterm/wezterm#4753] [Q&A]
### What Operating System(s) are you running on?

Windows

### WezTerm version

20231107-082518-f0e3eecb

### Ask your question!

I want to send code from neovim to IPython opened in second pane.
I cant send several lines using `wezterm cli send-text`
I want to try to do it this way:
- copy code to clipboard

create wezterm hotkey for this:
- activate pane by id
- paste text to pane from clipboard
- send enter key
- return to previous pane

I tried to create a hotkey using `wezterm.action_callback` but it didn't work out.
Could you help me with creating a hotkey?
Or suggest how else you can solve this problem?
Are there any other ways to transfer data between neovim and wezterm besides the clipboard?

**Comment by wez** [comment]
I'd suggest piping your multiple lines through `wezterm cli send-text --no-paste` on its stdin

**Reply by russarr** [reply]
I've dealt with sending to stdin(using lua io.popen)
Now I need to send "Enter" key to pane
There is no method "send-key" in wezterm cli. Any plans to this feature? 
As a solution, I want to try to make a hotkey that sends an enter key to the pane. Is there such a way?
Or you need to switch to the pane. Send a key and return to the previous pane?
If so, can you help to create a composite hotkey?

**Comment by russarr** [comment]
If it's not too much trouble, can you give me an example? I do not know how to do this.

**Reply by wez** [reply]
`echo -e "id\nid\n" | wezterm cli send-text --no-paste`

**Reply by wez** [reply]
that will run the `id` command twice in a row

**Comment by russarr** [comment]
Unfortunately, there is no "-e" option in echo command on Windows. I tried googling, but I didn't find a quick answer. I will try to search for information further.
As an interim solution, you can help with keymap for wezterm, which sends the contents of the clipboard to the pane by id?

**Reply by akthe-at** [reply]
> Unfortunately, there is no "-e" option in echo command on Windows. I tried googling, but I didn't find a quick answer. I will try to search for information further.
> As an interim solution, you can help with keymap for wezterm, which sends the contents of the clipboard to the pane by id?

Did you ever find a good fix and workflow for this? I am trying to figure out a good flow guy sending lines from neovim to a split and hitting enter without losing focus in neovim.

**Reply by aplatypus** [reply]
I do not recall having a problem with `echo` on windows, it was long ago though.  What you can always do on Windows is to `type text-to-send.txt | wezterm  cli send-text --no-paste ` .  As a kind of workaround.  You can probably ask Google how to send a CR/LF pair from `echo` on windows (and LF/CR for Mac-s).

**Comment by aplatypus** [comment]
I was looking for a way to send commands  to a shell on `Linux` and happily this works:

**Terminal 2**
 * This example injects a command line for the destination terminal window/tab.

```shell
 $ echo -e "echo \"Hi there...\"\n" | wz cli send-text --no-paste
```

**Terminal instance 1**
* An idle window and then ...
   * The sent `echo` command is executed.
  
```shell
 $
 $ echo "Hi there..."
Hi there...
 $
```

I love this because it gives me back the X.11 `xdotool`  command to send text and key strokes to `stdin` that was deprecated by Wayland.  The main use cases for this are:

* Automation or replay macros
* Testing on the command line
* Mirroring commands -- Terminator lets you group panels such that every keystroke is delivered to multiple sessions at the same time.

This example only lets one send stuff via `stdin`.  Fortunately that's sufficient for my own overdue tab auto-initialise script.  This is a big _**win**_ from my perspective.    I want this ability to remain. :-)

**Comment by aplatypus** [comment]
For a very long time, I have looked for a way to send (inject) commands  to a shell on `Linux` and happily this works:

**Terminal 2**
 * This example injects a command line for the destination terminal window/tab.

```shell
 $ echo -e "echo \"Hi there...\"\n" | wz cli send-text --no-paste
```

**Terminal instance 1**
* An idle window and then ...
   * The sent `echo` command is executed.
  
```shell
 $
 $ echo "Hi there..."
Hi there...
 $
```

I love this because it gives me back the X.11 [`xdotool`](https://github.com/jordansissel/xdotool#readme)  command to send text and key strokes to `stdin` that was deprecated by Wayland.  The main use cases for this might be:

* Automation or replay macros
* Testing on the command line
* Mirroring commands -- For example: [`Terminator`](https://en.wikipedia.org/wiki/Terminator_(terminal_emulator)) lets you group panels such that input is delivered to multiple sessions at the same time.

Even this simple send via `stdin` is sufficient for my overdue tab set-up script.   I'd vote is to retain this facility :-)

I am moved to comment that it would be a great boon to have something like [`xdotool`](https://github.com/jordansissel/xdotool#readme) for Wayland that operates similar to the `cli `**`send-text`**` --no-paste` techique.  Ideally that kind of tooling is better as a generic "helper" tool though, not tied to a specific application _imho_.

**Comment by ManuGraiph** [comment]
How can i define the target terminal/pane? I'm currently using tmux literally only for the send-text command, if i could do the same in pure wezterm, it'd be great!

**Reply by LazyRen** [reply]
You can run `wezterm cli list` to get full list of panes or simply run `wezterm cli get-pane-direction next` to get next pane.
Retrieved pane-id can then be used with `--pane-id` argument.

**Reply by aplatypus** [reply]
@ManuGraiph ... when you make a new pane WezTerm cli emits the pane id (on the command line).  If you use Ctrl/Shift-T it may show in the log file(???).
